lib/provider-manager.coffee

```diff
@@ -36,37 +36,58 @@ class ProviderManager
     @globalBlacklist = null
     @providers = null
 
-  providersForScopeDescriptor: (scopeDescriptor) =>
+  applicableProviders: (editor, scopeDescriptor) =>
+    processedProviders = @processProvidersByEditor(@providers, editor)
+    processedProviders = @processProvidersByScopeDescriptor(processedProviders, scopeDescriptor)
+    processedProviders = @processProvidersByPriority(processedProviders)
+    @removeMetadata(processedProviders)
+
+  processProvidersByScopeDescriptor: (providers, scopeDescriptor) =>
     scopeChain = scopeChainForScopeDescriptor(scopeDescriptor)
     return [] unless scopeChain
     return [] if @globalBlacklistSelectors? and selectorsMatchScopeChain(@globalBlacklistSelectors, scopeChain)
 
     matchingProviders = []
     disableDefaultProvider = false
-    lowestIncludedPriority = 0
-
-    for providerMetadata in @providers
+    defaultProviderMetadata = null
+    for providerMetadata in providers
       {provider} = providerMetadata
+      if provider is @defaultProvider
+        defaultProviderMetadata = providerMetadata
       if providerMetadata.matchesScopeChain(scopeChain)
-        matchingProviders.push(provider)
+        matchingProviders.push(providerMetadata)
         if provider.excludeLowerPriority
           lowestIncludedPriority = Math.max(lowestIncludedPriority, provider.inclusionPriority ? 0)
         if providerMetadata.shouldDisableDefaultProvider(scopeChain)
           disableDefaultProvider = true
 
     if disableDefaultProvider
-      index = matchingProviders.indexOf(@defaultProvider)
+      index = matchingProviders.indexOf(defaultProviderMetadata)
       matchingProviders.splice(index, 1) if index > -1
 
-    matchingProviders = (provider for provider in matchingProviders when (provider.inclusionPriority ? 0) >= lowestIncludedPriority)
-    stableSort matchingProviders, (providerA, providerB) =>
-      difference = (providerB.suggestionPriority ? 1) - (providerA.suggestionPriority ? 1)
+    stableSort matchingProviders, (providerA, providerB) ->
+      difference = (providerB.provider.suggestionPriority ? 1) - (providerA.provider.suggestionPriority ? 1)
       if difference is 0
-        specificityA = @metadataForProvider(providerA).getSpecificity(scopeChain)
-        specificityB = @metadataForProvider(providerB).getSpecificity(scopeChain)
+        specificityA = providerA.getSpecificity(scopeChain)
+        specificityB = providerB.getSpecificity(scopeChain)
         difference = specificityB - specificityA
       difference
 
+  processProvidersByEditor: (providers, editor) ->
```

Comment:
I guess the other ones are called "process". I wonder if we might come up with a more descriptive verb.